components/forms/src/api.rs

```diff
@@ -0,0 +1,553 @@
+            postal_code     = :postal_code,
+            country         = :country,
+            tel             = :tel,
+            email           = :email
+        WHERE guid          = :guid",
+        rusqlite::named_params! {
+            ":given_name": address.address_fields.given_name,
+            ":additional_name": address.address_fields.additional_name,
+            ":family_name": address.address_fields.family_name,
+            ":organization": address.address_fields.organization,
+            ":street_address": address.address_fields.street_address,
+            ":address_level3": address.address_fields.address_level3,
+            ":address_level2": address.address_fields.address_level2,
+            ":address_level1": address.address_fields.address_level1,
+            ":postal_code": address.address_fields.postal_code,
+            ":country": address.address_fields.country,
+            ":tel": address.address_fields.tel,
+            ":email": address.address_fields.email,
+            ":guid": address.guid,
+        },
+    )?;
+
+    tx.commit()?;
+    Ok(())
+}
+
+#[allow(dead_code)]
+fn delete_address(conn: &mut Connection, guid: &Guid) -> Result<bool> {
+    let tx = conn.transaction()?;
+
+    // check that guid exists
+    let exists = tx.query_row(
+        "SELECT EXISTS (
+            SELECT 1
+            FROM addresses_data d
+            WHERE guid = :guid
+                AND NOT EXISTS (
+                    SELECT 1
+                    FROM   addresses_tombstones t
+                    WHERE  d.guid = t.guid
+                )
+        )",
+        &[guid.as_str()],
+        |row| row.get(0),
+    )?;
+
+    if exists {
+        tx.execute_named(
+            "INSERT OR IGNORE INTO addresses_tombstones (
+                guid,
+                time_deleted
+            ) VALUES (
+                :guid,
+                :time_deleted
+            )",
+            rusqlite::named_params! {
+                ":guid": guid.as_str(),
+                ":time_deleted": get_current_time_ms(),
+            },
+        )?;
+    }
+
+    tx.commit()?;
+    Ok(exists)
+}
+
+fn get_current_time_ms() -> i64 {
+    let d = time::SystemTime::now()
+        .duration_since(time::UNIX_EPOCH)
+        .unwrap_or_default();
+
+    (d.as_secs() as i64) * 1000 + (i64::from(d.subsec_nanos()) / 1_000_000)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::db::test::new_mem_db;
+
+    #[test]
+    fn test_address_create_and_read() {
+        let mut db = new_mem_db();
+
+        let saved_address = add_address(
+            &mut db,
+            NewAddressFields {
+                given_name: "jane".to_string(),
+                family_name: "doe".to_string(),
+                street_address: "123 Main Street".to_string(),
+                address_level2: "Seattle, WA".to_string(),
+                country: "United States".to_string(),
+
+                ..NewAddressFields::default()
```

Comment:
Nothing to do with anything, but I love Rust's default initialization syntax! 😍🦀